File: 396-rotate-function/rotate-function.py
class Solution:
    def maxRotateFunction(self, nums: List[int]) -> int:
        total_sum = sum(nums)
        n = len(nums)
        current_f = sum(i * num for i, num in enumerate(nums))
        max_f = current_f
        
        for i in range(1, n):
            current_f += total_sum - n * nums[n - i]
            if current_f > max_f:
                max_f = current_f
                
        return max_f

File: 396-rotate-function/rotate-function.java
class Solution {
    public int maxRotateFunction(int[] nums) {
        int totalSum = 0;
        int n = nums.length;
        int currentF = 0;
        
        for (int i = 0; i < n; i++) {
            totalSum += nums[i];
            currentF += i * nums[i];
        }
        
        int maxF = currentF;
        
        for (int i = 1; i < n; i++) {
            currentF += totalSum - n * nums[n - i];
            if (currentF > maxF) {
                maxF = currentF;
            }
        }
        
        return maxF;
    }
}